File: 코드/ch5/clip5/boj10971/main.py
import sys

def travel(start, node, total_cost, count):
    global answer
    if count == n and start == node:
        answer = min(answer, total_cost)
        return
    for i in range(n):
        if w[node][i] != 0 and not visited[i]:
            visited[i] = True
            travel(start, i, total_cost + w[node][i], count + 1)
            visited[i] = False

n = int(sys.stdin.readline())
w = []
for _ in range(n):
    w.append(list(map(int, sys.stdin.readline().split())))

visited = [False] * n
answer = float('inf')
travel(0, 0, 0, 0)

print(answer)

File: 코드/ch5/clip5/boj10971/Main.java
import java.util.Scanner;

class Main {
    static int n;
    static int[][] w;
    static boolean[] visited;
    static int answer = Integer.MAX_VALUE;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        w = new int[n][n];
        visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                w[i][j] = sc.nextInt();
            }
        }
        travel(0, 0, 0, 0);
        System.out.println(answer);
    }

    static void travel(int start, int node, int sum, int cnt) {
        // base case
        if (cnt == n && start == node) {
            answer = Math.min(sum, answer);
            return;
        }
        // recursive case
        for (int i = 0; i < n; i++) {
            if(!visited[i] && w[node][i] != 0){
                visited[i] = true;
                travel(start, i, sum + w[node][i], cnt + 1);
                visited[i] = false;
            }
        }
    }
}
